src/main/java/com/hierynomus/smbj/transport/tcp/AsyncDirectTcpTransport.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.transport.tcp;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.AsynchronousChannelGroup;
+import java.nio.channels.AsynchronousSocketChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hierynomus.protocol.Packet;
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+import com.hierynomus.smbj.lock.CrossThreadLock;
+import com.hierynomus.smbj.transport.PacketHandlers;
+import com.hierynomus.smbj.transport.TransportException;
+import com.hierynomus.smbj.transport.TransportLayer;
+
+/**
+ * A transport layer over Direct TCP/IP that uses asynchronous I/O.
+ */
+public class AsyncDirectTcpTransport<P extends Packet<P, ?>> implements TransportLayer<P> {
+    public static final int DEFAULT_CONNECT_TIMEOUT = 5000;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final PacketHandlers<P> handlers;
+    private final AsynchronousSocketChannel socketChannel;
+    private final AsyncPacketReader<P> packetReader;
+    private int soTimeout = 0;
+
+    // AsynchronousSocketChannel doesn't support concurrent writes
```

Comment:
I think you mean this section from the JavaDoc

> Channels of this type are safe for use by multiple concurrent threads. They support concurrent reading and writing, though at most one read operation and one write operation can be outstanding at any time. If a thread initiates a read operation before a previous read operation has completed then a ReadPendingException will be thrown. Similarly, an attempt to initiate a write operation before a previous write has completed will throw a WritePendingException.

The reason why this is the case is that having 2 writes to a socket means that you have no clue how the bytes are interleaved. You must finish 1 write before you can write the next, else you get garbage on the line. This does mean you need to have external synchronization or a queue indeed. But that should be handled on the transport level. I really don't understand your option 1.
Option 2 makes life simple, and could work very well, for instance using a `LinkedBlockingQueue`. I think it isn't more complex, but maybe even simpler
For option 3, I still don't understand why you need a 'CrossThreadLock'. What is the reasoning for not using a regular ReentrantLock?